### PR TITLE
feat: add browsers list, linting plugin and Docs section with supported browsers

### PR DESCRIPTION
**Pull request by oemueller**
Relates to #83 

## Checklist

- [x] The added / edited code has been documented with [JSDoc](https://jsdoc.app/about-getting-started)
- [x] A changeset is added with `npx changeset add` if your changes should be released as npm package (because they affect the library usage)

### CODE REVIEW

**Comment by oemueller** [comment]
Preview of browser section on Onyx Docs:
Bildschirmfoto 2024-10-15 um 19 10 21

File: .changeset/brave-days-happen.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+---
+"docs": minor
+"sit-onyx": minor
```
Comment (larsrickert) [inline .changeset/brave-days-happen.md:3]:
```suggestion
```

There are no relevant changes for the users, right? So we don't need to publish a new onyx version

File: .eslintrc.cjs [hunk 2]
```diff
@@ -18,6 +22,7 @@ module.exports = {
     // see: https://github.com/vuejs/eslint-config-typescript#vueeslint-config-typescriptrecommended
     "@vue/eslint-config-typescript/recommended",
     "@vue/eslint-config-prettier/skip-formatting",
+    "plugin:compat/recommended",
```
Comment (larsrickert) [inline .eslintrc.cjs:25]: Is this the line that will check all our code to match the browserlistrc?

File: apps/docs/package.json [hunk 3]
```diff
@@ -26,5 +27,11 @@
     "vitepress": "^1.4.0",
     "vue": "catalog:",
     "vue-chartjs": "^5.3.1"
+  },
+  "imports": {
+    "#src/.vitepress/browser-loader.data": {
+      "storybook": "./src/.vitepress/browser-loader.data.mock.ts",
+      "default": "./src/.vitepress/browser-loader.data.ts"
+    }
```
Comment (larsrickert) [inline apps/docs/package.json:35]: Why is this needed?

File: apps/docs/src/.vitepress/browser-loader.data.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+import path, { dirname } from "path";
+import { fileURLToPath } from "url";
+import { defineLoader } from "vitepress";
+
+const __filename = fileURLToPath(import.meta.url);
+const __dirname = dirname(__filename);
+const browserslistRcPath = path.resolve(__dirname, "../../../../.browserslistrc");
+
+const load = async () => {};
+
```
Comment (larsrickert) [inline apps/docs/src/.vitepress/browser-loader.data.ts:12]:
```suggestion
```

This is obsolete, right?

File: apps/docs/src/.vitepress/browser-loader.data.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
+import fs from "fs";
+import https from "https";
+import path, { dirname } from "path";
+import { fileURLToPath } from "url";
+import { defineLoader } from "vitepress";
+
+const __filename = fileURLToPath(import.meta.url);
+const __dirname = dirname(__filename);
+const browserslistRcPath = path.resolve(__dirname, "../../../../.browserslistrc");
```
Comment (larsrickert) [inline apps/docs/src/.vitepress/browser-loader.data.ts:9]:
```suggestion
const browserslistRcPath = fileURLToPath(new URL("../../../../.browserslistrc", import.meta.url));
```

Usually we do this like this so we save 2 lines :)

File: apps/docs/src/.vitepress/browser-loader.data.ts [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+  coverage: number;
+  id: string;
+  name: string;
+  versions: Record<string, number | undefined>;
+};
+
+export interface Data {
+  browsers: Browser[];
+  coverage: number;
+  config: string;
```
Comment (larsrickert) [inline apps/docs/src/.vitepress/browser-loader.data.ts:23]:
```suggestion
```

Is config unused?

File: apps/docs/src/.vitepress/browser-loader.data.ts [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+ * More information can be found on https://browsersl.ist
+ */
+export default defineLoader({
+  async load(): Promise<Data> {
+    return new Promise((resolve, reject) => {
+      let browserRules = "";
+
+      try {
+        const data = fs.readFileSync(browserslistRcPath, "utf8");
+        const [_firstLine, ...lines] = data.split("\n").filter((l) => !!l);
```
Comment (larsrickert) [inline apps/docs/src/.vitepress/browser-loader.data.ts:42]: Omitting the first line might not work as expected if comments are removed/added to the file

Should we instead remove all lines that are comments (start with a a `#`) to be safer?

File: apps/docs/src/.vitepress/browser-loader.data.ts [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+        browserRules = lines.join("").trim();
+      } catch (e) {
+        console.error("could not read .browserslistrc");
+      }
+
+      const url = `https://browsersl.ist/api/browsers?q=${browserRules}`;
+
+      https
```
Comment (larsrickert) [inline apps/docs/src/.vitepress/browser-loader.data.ts:50]: Is there a reason you are using https instead if the default fetch?

File: apps/docs/vite.config.ts [hunk 9]
```diff
@@ -5,6 +5,7 @@ import { VITE_BASE_CONFIG } from "./node_modules/@sit-onyx/shared/vite.config.ba
 // https://vitejs.dev/config
 export default defineConfig({
   ...VITE_BASE_CONFIG,
+  optimizeDeps: { exclude: ["fsevents"] },
```
Comment (larsrickert) [inline apps/docs/vite.config.ts:8]: Why is this needed? Removing it also works fine for me :)

File: apps/docs/src/.vitepress/components/OnyxHomePage.vue
Comment (larsrickert) [inline apps/docs/src/.vitepress/components/OnyxHomePage.vue:1]: I appreciate your detailed approach here 👍 
But at least for me, all the coverage and browser version numbers are a bit overwhelming. In fact, what does the browser coverage actually mean?^^ That we only support 44.33% or that 44.33% of users are using Chrome on Android?

tl;rd: I have 2 sugesstions:
1. Maybe you simplify the browser support to something like this + and additional note that onyx might work with older versions but it is not tested/guaranteed: image
2. We move the component to the [getting started page](https://onyx.schwarz/development) since it seems more fitting to me there

What do you think?

**Review by larsrickert — CHANGES_REQUESTED** [review]
Thanks for your contribution! I left a few comments, let me know what you think!

**Comment by larsrickert** [comment]
With the current approach, only JavaScript features are linted, right? So if we use CSS features that are e.g. not supported in firefox, we will not know

File: .eslintrc.cjs (hunk 2 above)
Comment (oemueller) [reply .eslintrc.cjs:25]: that's the plugin to check our code, runs as part of `lint:ci:all`

File: apps/docs/package.json (hunk 3 above)
Comment (oemueller) [reply apps/docs/package.json:35]: because the file normally runs in Node.js environment and this would throw an error in Storybook context. 
This is the Storybook recommended way to mock the import.

https://vitepress.dev/guide/data-loading

File: apps/docs/src/.vitepress/browser-loader.data.ts (hunk 8 above)
Comment (oemueller) [reply apps/docs/src/.vitepress/browser-loader.data.ts:46]: will check :-)

File: apps/docs/src/.vitepress/components/OnyxHomePage.vue
Comment (oemueller) [reply apps/docs/src/.vitepress/components/OnyxHomePage.vue:1]: makes sense, shall we include mobile browsers as well ?

**Comment by oemueller** [comment]
we could try out https://stylelint.io/ to lint our CSS, will create a follow up ticket for that

File: apps/docs/src/.vitepress/components/OnyxHomePage.vue
Comment (larsrickert) [reply apps/docs/src/.vitepress/components/OnyxHomePage.vue:1]: Sure why not :)

File: apps/docs/src/.vitepress/components/BrowsersList.vue [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+const filteredBrowsers = popularBrowsers
+  .map((pb) => {
+    const b = browsers.find((b) => b.id === pb);
+    if (b)
+      return {
+        ...b,
+        minVersion: b?.versions ? Object.keys(b.versions).sort()[0] : "",
+      };
+  })
+  .filter((b) => !!b);
```
Comment (larsrickert) [inline apps/docs/src/.vitepress/components/BrowsersList.vue:17]:
```suggestion
const filteredBrowsers = browsers
  .filter((browser) => popularBrowsers.includes(browser.id))
  .map((browser) => {
    return {
      ...browser,
      minVersion: Object.keys(browser.versions).sort()[0],
    };
  });
```

File: apps/docs/src/.vitepress/components/BrowsersList.vue [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+      <p>Version ≥ {{ minVersion }}</p>
+    </li>
+  </ul>
+</template>
+
+<style lang="scss" scoped>
+.browsersList {
+  display: grid;
+  grid-template-columns: repeat(auto-fill, minmax(160px, 1fr));
+  row-gap: var(--onyx-spacing-lg);
```
Comment (larsrickert) [inline apps/docs/src/.vitepress/components/BrowsersList.vue:40]:
```suggestion
  display: flex;
  flex-wrap: wrap;
  gap: var(--onyx-spacing-xl);
```

We could use flex + flex-wrap instead so we don't need to maintain the 160px width manually and instead let the components take whatever width they need :) What do you think?

File: apps/docs/src/.vitepress/components/BrowsersList.vue [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+</script>
+
+<template>
+  <ul class="browsersList">
+    <li v-for="{ name, minVersion, id } in filteredBrowsers" :key="name" class="browser">
+      <img
+        :src="`/assets/browsers/${id}.svg`"
+        :alt="`${name} icon`"
+        width="40px"
+        height="40px"
```
Comment (larsrickert) [inline apps/docs/src/.vitepress/components/BrowsersList.vue:27]: Should we (also) set the size via CSS? Otherwise the images are not exactly 40x40px:
image

File: apps/docs/src/.vitepress/components/BrowsersList.vue [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+    line-height: 1.3;
+  }
+  &__name {
+    padding-top: var(--onyx-spacing-xs);
+    font-weight: 600;
+  }
+  &__image {
+    position: relative;
+    transform: translateX(-4px);
+  }
```
Comment (larsrickert) [inline apps/docs/src/.vitepress/components/BrowsersList.vue:61]: Might be cleaner to remove the whitespace from the images itself, what do you think?
I can provide you the edited images

File: apps/docs/src/assets/browsers/chrome.svg
Comment (larsrickert) [inline apps/docs/src/assets/browsers/chrome.svg:1]: Where did you get the browser logos from? Are they copyright free / free to use?

File: apps/docs/src/development/index.md [hunk 14]
```diff
@@ -154,8 +154,24 @@ Be careful when overriding styles globally since it will affect **EVERY** compon
 
 :::
 
+### Browser Support
```
Comment (larsrickert) [inline apps/docs/src/development/index.md:157]: Would prefer to move this to the bottom of the page and change the headline to h2 :)

File: apps/docs/src/.vitepress/components/BrowsersList.vue
Comment (larsrickert) [inline apps/docs/src/.vitepress/components/BrowsersList.vue:1]: I like the new design, good job!

File: apps/docs/src/development/index.md [hunk 15]
```diff
@@ -154,8 +154,24 @@ Be careful when overriding styles globally since it will affect **EVERY** compon
...
+import BrowsersList from "../.vitepress/components/BrowsersList.vue"
+</script>
+
+This list of supported browsers is based on our `Browserslist` config and data from [caniuse](https://caniuse.com)
+
+Onyx might also work in older versions, but we cannot guarantee full compatibility.
+
+<BrowsersList />
+
+Find our .browserslistrc [here](https://github.com/SchwarzIT/onyx/blob/main/.browserslistrc)
```
Comment (larsrickert) [inline apps/docs/src/development/index.md:169]:
```suggestion
## Browser Support

<script lang="ts" setup>
import BrowsersList from "../.vitepress/components/BrowsersList.vue"
</script>

Onyx works best with the following browser versions. Older versions might also work but we can't guarantee full compatibility.

<BrowsersList />

::: details More information (.browserslistrc file)

<<< @include ../../../../.browserslistrc
:::
```

What do you think about rephrasing it like this and include the browserlistrc content directly?

File: apps/docs/src/development/index.md
Comment (larsrickert) [inline apps/docs/src/development/index.md:1]: @JoCa96 If we list supported browsers in our docs, we should also make sure that they are actually working, so e.g. use those browsers and versions in our Playwright tests. What do you think?

File: apps/docs/src/assets/browsers/chrome.svg
Comment (oemueller) [reply apps/docs/src/assets/browsers/chrome.svg:1]: they are from https://github.com/browserslist/browsersl.ist which is under MIT license so should be fine

File: apps/docs/src/development/index.md (hunk 14 above)
Comment (oemueller) [reply apps/docs/src/development/index.md]: like it!

File: apps/docs/src/.vitepress/components/BrowsersList.vue (hunk 13 above)
Comment (oemueller) [reply apps/docs/src/.vitepress/components/BrowsersList.vue]: sure, fill free to change them :-)

File: .eslintrc.cjs
[the remark this answers is not in the document]
Comment (oemueller) [reply .eslintrc.cjs:1]: fixed + rebased :-)

File: apps/docs/src/development/index.md [hunk 16]
```diff
@@ -154,8 +154,25 @@ Be careful when overriding styles globally since it will affect **EVERY** compon
...
+<<< @include ../../../../.browserslistrc{txt}
+:::
+
+## Next steps
 
 - [Theming](/development/theming), e.g. for setting up the dark mode
 - [i18n](/development/i18n) if you are supporting multiple languages in your application
 - [Grid](/development/grid) for easily implementing custom layouts
+
+---
```
Comment (larsrickert) [inline apps/docs/src/development/index.md:178]:
```suggestion
```

Is this a leftover or intended? :)

**Review by larsrickert — APPROVED** [review]

File: apps/docs/src/development/index.md (hunk 16 above)
Comment (oemueller) [reply apps/docs/src/development/index.md]: leftover 😄

File: apps/docs/src/development/index.md
Comment (oemueller) [reply apps/docs/src/development/index.md:1]: shall we create follow up ticket ?

Comment (larsrickert) [reply apps/docs/src/development/index.md:1]: #1997

I will refine it next week

**Review by larsrickert — APPROVED** [review]
